### ISSUE TITLE
Security fix breaks `guake -e`

### ISSUE THREAD

**Issue by midnight-wonderer** [Guake/guake#2042]
In version 3.8.5, `execute_command` has been removed from DBus
https://github.com/Guake/guake/issues/1796

The discussion there implies that `-e` option won't affect by this change.
In reality, it does break the functionality

Here the stack trace
~~~
Traceback (most recent call last):
  File "$HOME/.local/bin/guake", line 8, in <module>
    sys.exit(exec_main())
  File "$HOME/.local/lib/python3.10/site-packages/guake/main.py", line 633, in exec_main
    if not main():
  File "$HOME/.local/lib/python3.10/site-packages/guake/main.py", line 552, in main
    remote_object.execute_command(options.command)
  File "$HOME/.local/lib/python3.10/site-packages/dbus/proxies.py", line 72, in __call__
    return self._proxy_method(*args, **keywords)
  File "$HOME/.local/lib/python3.10/site-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File "$HOME/.local/lib/python3.10/site-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Traceback (most recent call last):
  File "$HOME/.local/lib/python3.10/site-packages/dbus/service.py", line 662, in _message_cb
    (candidate_method, parent_method) = _method_lookup(self, method_name, interface_name)
  File "$HOME/.local/lib/python3.10/site-packages/dbus/service.py", line 254, in _method_lookup
    raise UnknownMethodException('%s is not a valid method' % method_name)
dbus.exceptions.UnknownMethodException: org.freedesktop.DBus.Error.UnknownMethod: Unknown method: execute_command is not a valid method
~~~

To reproduce, simply try `-e` option.

$guake --support
~~~
<details><summary>$ guake --support</summary>

Guake Version:		3.8.5

Vte Version:		0.60.3

Vte Runtime Version:	0.60.3

--------------------------------------------------
GTK+ Version:		3.24.20

GDK Backend:		<GdkX11.X11Display

--------------------------------------------------
Desktop Session: gnome

--------------------------------------------------
Display: :0

RGBA visual: True

Composited: True

* Monitor: 0 - XWAYLAND0
    * Geometry:		1920 x 1080 at 0, 0
    * Size:		480 x 270 mm²
    * Primary:		True
    * Refresh rate:	59.962 Hz
    * Subpixel layout:	unknown
~~~

**Comment by midnight-wonderer** [comment]
Not sure how D-Bus work, but can we somehow limit D-Bus access only to the command from the same user running guake,  and restore the functionality?

**Comment by Davidy22** [comment]
Hrm, when Guake is run via terminal it's always going to be a separate instance from the one you want to interact with. It contacts the current running Guake instance via dbus by name, and then it can send commands. Somehow, we need to generate a secret that only a running instance of Guake and another instance of Guake can access, a random secret of some kind might be nice if there's a way to only make it viewable to instances of Guake. Not sure how we can pass the secret between the two instances in a way that can't be replicated by other programs.

Actually, the exploit that I was alerted to and dealt with would allow malicious actors to gain the ability to execute commands as root if Guake is in a super user state. Since the the goal of the exploit is to gain root, if there's a secret stored in a location only accessible with super user privileges, I think that deals with the issue. Most people aren't running Guake with sudo normally, so it wouldn't normally have the ability to access information in a place with root privileges so I think there'll need to be a one-time setup of a public/private key pair. So the way it'd go is:

- Prior setup - user goes to preferences, clicks on button to enable -e, enters password and Guake generates and stores a private key in root only place and public key where it can see it normally

- When invoking -e - Attempts to view private key, fails if not invoked with sudo. Otherwise uses private key to authenticate the dbus call.

Have I missed an angle of attack here? I can run this by the person who alerted us to the initial issue as well, since they were savvy enough to spot the original vulnerability.

**Comment by midnight-wonderer** [comment]
If we are going that route, why not generate the secret on the first run?  
It seems much simpler.

For instance, store it as a flat file at `~/.guake/dbus-key` and set the permission to 0600.  
This way we can guarantee that the user can only interact with guake instances run by themselves.

**Comment by Davidy22** [comment]
Hum, is that sufficient? I'll include that in the message I send the security guy. Wouldn't this allow another installed program running at user level to perform the exploit?

**Comment by midnight-wonderer** [comment]
If they have access to the file with 0600 permission, they have no reason to try running a command as that user; they already have higher privileges.

I'm glad we can discuss solutions and hope we can come up with a good resolution.
Thank you for the follow-up.

**Comment by ivoshm** [comment]
I'm not a D-Bus expert, but if I remember correctly, there are two type of D-Buses - one system D-Bus and more session D-Buses, which are only accessible in a given individual user login sessions.
I would expect Guake to use exclusively "session D-Bus" and then the implied attack vector isn't so critical.

Excerpt from Wikipedia: https://en.wikipedia.org/wiki/D-Bus
... In practice, this means that any user process can connect to the system bus and to its current session bus, but not to another user's session buses, or even to a different session bus owned by the same user. ...

**Comment by Davidy22** [comment]
Still waiting on the reply from the security guy and was a bit busy last few days, can take a shot at this while waiting maybe. Going to try get it checked more than usual, really don't want to accidentally reintroduce the issue.

**Comment by Davidy22** [comment]
Initial implementation with notes in #2051. File set with 400 permissions post creation, exploit that was initially reported still possible, may need revision.

**Comment by The-Plottwist** [comment]
Any news?

By the way, how does other terminals like `gnome-terminal` or `xfce4-terminal` implement this feature?
Maybe they have a solution about this.

**Comment by ivoshm** [comment]
Sorry, I've studied the proposed solution a bit and I feel that it doesn't solve the problem. 

After all, the basic problem is that the background guake process that handles D-BUS calls is (stupidly) running as root.

Isn't it easier to just add to original **dbusiface.py** something like this:

```
   def execute_command(self, command):
     if os.geteuid() == 0:
          sys.stdout.write(" Guake is running as root, unable to use execute_command")
        else:
          self.guake.execute_command(command)     
```
(sorry - I don't speak Pythonicano :) )

Because if I'm regular (non-root) user and my guake is running under my effective id, there is IMHO no security limitation to use execute_command via DBUS call and there is no need to mess with "sudo" and "cryptography".

**Comment by Davidy22** [comment]
Actually, Guake can execute root commands even when it isn't running as root. As a terminal, a Guake session that isn't running as root can be escalated to root privileges by typing sudo, which you likely do often, or by entering `su -`. Both of these will put Guake into a state where a password is no longer needed to execute a command that would normally require root privileges, though the first one only temporarily and only under some admittedly common configurations. os.geteuid() will tell us if Guake was started with `sudo guake` but then if the user types sudo once and enters the five minute period where the terminal stops asking you to keep typing your password os.geteuid() will continue to report that the user is not root even when the exploit is possible.

**Comment by ivoshm** [comment]
> @ivoshm Root escalation is only a part of the issue. Even if you start your guake terminal as a non-root user, other users can run commands as you. e.g., Someone else can steal your key by sending this: `cp ~/.ssh/id_ed25519 /tmp && chmod 0666 /tmp/id_ed25519` via dbus.

That's not true (at least) on my Manjaro - Guake use **session D-BUS** which is connected to one (my) user session in graphical environment. Other users (except "root" of course) can't connect to it.

Example with Guake 3.8.1 when **user1001** try to send command to **user1000** session D-BUS
`
user1001$ DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus gdbus call -e -d org.guake3.RemoteControl -o /org/guake3/RemoteControl -m org.guake3.RemoteControl.execute_command "id > /tmp/guake_exec_id.txt"
Error connecting: Could not connect: Permission denied
`

**Comment by midnight-wonderer** [comment]
Good to know D-BUS has something in place already. Then, this is just a misunderstanding on my part.
I obviously never worked with D-BUS, so I assumed the worst and proposed based on the assumption that D-BUS is similar to TCP.
Then, I can't comment on your solution because it needs a deeper understanding of D-BUS.

More importantly, can we step back a little?
After reading David's scenario, I don't feel confident in this anymore; it is getting messier.

Is it only me, or is everyone here just to configure the startup tabs?  
I mean this thing:
![Screenshot](https://i.imgur.com/zdMQJl1.png)
I just want a tab on guake start where I can control my AirCon via shell. So I simply configure it to standby for my command.
~~~bash
guake -r "AirCon Controller" -e "netcat -u 192.168.<AirCon IP Address> <Control Port>"
~~~
I don't actually care if I can remote control guake or not. I am fine with this feature gone forever if I can config startup tabs by other means.
Can someone tell me? Why on earth do you want to remote control the terminal apart from the use case I described.
We might be better off removing the remote controlling feature after all.

**Comment by Davidy22** [comment]
Checking ID catches if the user in `su -` but I don't believe it detects the temporary state where a terminal remembers the user's sudo password and stops asking for it if you entered it in the last five minutes.

If the -e command was being used solely for startup scripts, what about reforming it to always run the command in a new tab? I think that dodges any escalated privilege hijacking concerns, just need a moment to try and brainstorm any way a fresh shell could still have root.

**Comment by Davidy22** [comment]
I couldn't think of any way to hijack privileges on a fresh terminal so I'm open to going that way too. I'll put it to a vote, all in favor of reforming -e to only make a fresh tab to run in leave a spaceship reaction, all in favor of making -e sudo leave an eyes emoji

**Comment by midnight-wonderer** [comment]
I was thinking something along the line of dedicated config sections for startup tabs.
But, I don't contribute code, so I don't have much say in it, naturally.

If there needs to be something between these two, the rocket one works for me.
Added that I want to be able to create a new tab with a startup command and also set the name of the tab.

**Comment by Davidy22** [comment]
New approach created, was significantly simpler, still want to see it get reviewed, @mlouielu, just to make sure I haven't gone and reintroduced a hole somewhere

**Comment by The-Plottwist** [comment]
Oh, I am relieved that this is finally settled.

It has been a nice discussion.

Good work everyone, cya.

**Comment by leomayer** [comment]
Any updates on this issue? Since its still not working :-(

**Comment by Davidy22** [comment]
Merged the resolution after far too much waiting. I'm just going to stop requesting reviews since other reviewers are fully absent now

**Comment by vasilakisfil** [comment]
Can we bring back the original functionality somehow? Even if that means that you have to use a token or an authentication, whatever I don't know, but I think we cut down too much for security here. I think security is paramount, but the solution is not to remove features but to make them secure.

I used to use execute_command to run specific commands on my guake tabs from vim. Is there any way to re-introduce the initial behavior, even behind an option or with the use of a security token or whatever (I have no idea of the d-bus internals).

**Comment by midnight-wonderer** [comment]
@vasilakisfil  
I don't know if it helps, but to recap the highlight of the discussion:

The vulnerability initially focused on dBus; however, later on, it was determined that letting 3rd party's application pose as if they are users themself is a problem; these include exposing an interface with the same effect as a user typing into the terminal.

Even with non-root tabs, the `sudo` seems to remember the password for a period of time.  
Of course, it is just a vulnerability, not exploitation; we simply don't know what we don't know.  
If you caught that part of the discussion already, my apology for the interruption.

**Comment by cmalard** [comment]
Still having the issue, did I missed something? ^^
```
Traceback (most recent call last):
  File "/usr/bin/guake", line 33, in <module>
    sys.exit(load_entry_point('guake==3.9.0', 'console_scripts', 'guake')())
  File "/usr/lib/python3/dist-packages/guake/main.py", line 648, in exec_main
    if not main():
  File "/usr/lib/python3/dist-packages/guake/main.py", line 567, in main
    remote_object.execute_command(options.command)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 72, in __call__
    return self._proxy_method(*args, **keywords)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.UnknownMethod: Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/dbus/service.py", line 662, in _message_cb
    (candidate_method, parent_method) = _method_lookup(self, method_name, interface_name)
  File "/usr/lib/python3/dist-packages/dbus/service.py", line 254, in _method_lookup
    raise UnknownMethodException('%s is not a valid method' % method_name)
dbus.exceptions.UnknownMethodException: org.freedesktop.DBus.Error.UnknownMethod: Unknown method: execute_command is not a valid method
```

<details><summary>$ guake --support</summary>

Guake Version:		3.9.0

Vte Version:		0.68.0

Vte Runtime Version:	0.68.0

--------------------------------------------------
GTK+ Version:		3.24.33

GDK Backend:		<GdkX11.X11Display

--------------------------------------------------
Desktop Session: ubuntu-xorg

--------------------------------------------------
Display: :0

**Comment by Davidy22** [comment]
3.9 should be it, is there an additional older version of guake installed running that's receiving the commands?

**Comment by orazdow** [comment]
Does this issue mean you cannot start guake from another command in a specific directory?

**Comment by besworks** [comment]
> Does this issue mean you cannot start guake from another command in a specific directory?

It does and this breaks integration with **Open Terminal Here...** and **Run in Terminal...** file manager features. I came up with a workaround :

```bash
#!/bin/sh

if [ $# -lt 2 ]; then
  guake --show --new-tab "$PWD";
  exit 0;
fi

while getopts "e:" flag; do
  case $flag in
    e) dir="$PWD";
      cmd="$OPTARG";
      [ -f "$OPTARG" ] &&
      dir="$(dirname $(realpath $OPTARG))" &&
      cmd="./$(basename $OPTARG)";
      guake --show -e "cd \"$dir\" && clear && $cmd";
      break
    ;;
    \?) exit 1
    ;;
  esac
done
```

This script is specifically for use with the above mentioned file manager functions (btw, to get this working in Nemo File Manager, you have to symlink `guake-wrapper` to `/usr/local/bin/alacritty` or some other [**known terminal**](https://github.com/linuxmint/nemo/issues/3483)).

If you just want to exec in a new tab from another guake terminal it wouldn't be desirable to manipulate the command like I've done above so this is what I've come up with:

```bash
#!/bin/sh

# get args as an array
test -n "$*" && args=("$@")

# initialize variables
cmd=""
dir=""
ext=()
add=""
i=0

# loop through args and extract the command and directory
while [ $i -lt ${#args[@]} ]; do
  case "${args[$i]}" in
    -e)
      # -e should always be followed by the command
      cmd="${args[$i+1]}"
      i=$((i+2))
      ;;
    --execute-command|--execute-command=*)
      if [[ "${args[$i]}" =~ ^--execute-command=.* ]]; then
        # If the current argument contains '=', extract the command part after '='
        cmd="${args[$i]#--execute-command=}"
        i=$((i+1))
      else
        # Else, the next argument is the command
        cmd="${args[$i+1]}"
        i=$((i+2))
      fi
      ;;
    -n|--new-tab)
      # -n or --new-tab can be followed by a directory
      dir="${args[$i+1]}"
      i=$((i+2))
      ;;
    *)
      # collect any exta arguments
      ext+=("${args[$i]}")
      i=$((i+1))
      ;;
  esac
done


if [ -z "$cmd" ]; then
  # if -e or --execute-command arg was not set pass all args to guake
  cmd="${args[@]}"
  guake "$cmd"
  exit
fi

# combine the remaining args
if [ ${#ext[@]} -gt 0 ]; then
  add="${ext[*]}"
fi

if [[ -n "$dir" && -d "$dir" ]]; then
  # if a valid directory is specified, execute command inside it
  guake --show -e "cd \"$dir\" && $cmd" "$add"
else
  # otherwise just run the command in the current directory
  guake --show -e "$cmd" "$add"
fi

```
